Customize tippy top to accept Top List file and icon's minimum resolution from user

The script currently works:

1. Only with [Alexa top sites](https://github.com/mozilla/tippy-top-sites/blob/master/make_manifest.py#L20) and
2. Returns icons only of there is one that is at least [96 pixel wide](https://github.com/mozilla/tippy-top-sites/blob/master/make_manifest.py#L124)

 
It would be useful if we could allow users to specify a top sites file (e.g. a csv file) and minimum resolution. This can greatly increase the re-usability of this script (https://mozilla-hub.atlassian.net/browse/DENG-407 is one of the examples).

I propose to add these 2 as cli options. Both of these options will be optional. If any of them is not specified, the script will work as before i.e. with Alexa top sites and 96 pixel width to be backward compatible.

If we want to be a bit obnoxious, we could test this in the unit tests (but means we'd almost always certainly fail the CI tests (unless we skillfully exclude the 2-4 specific domains above):

```js
const fetch = require("node-fetch");

...

it("should be scrapeable by embedly", (done) => {
  fetchSite(site.url)
    .then((data) => {
      var res = data.urls[site.url];
      assert.isUndefined(res.error_message);
      // or `assert.notEqual(res.type, "error", res.error_message);`
      done();
    }).catch(done);
});

...

function fetchSite(url) {
  return fetch(EMBEDLY_PROXY_URL, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      urls: [url]
    })
  }).then((res) => res.json());
}
```

#### Output:
```
  656 passing (22s)
  4 failing

  1) top_sites #38 Groupon should be scrapeable by embedly:
     AssertionError: expected 'HTTP 403: Forbidden' to equal undefined
      at Function.assert.isUndefined (node_modules/chai/lib/chai/interface/assert.js:478:32)
      at test/index.test.js:35:14

  2) top_sites #51 Lowes should be scrapeable by embedly:
     AssertionError: expected 'HTTP 403: Forbidden' to equal undefined
      at Function.assert.isUndefined (node_modules/chai/lib/chai/interface/assert.js:478:32)
      at test/index.test.js:35:14

  3) top_sites #64 Pinterest should be scrapeable by embedly:
     AssertionError: expected 'HTTP 550: Unknown' to equal undefined
      at Function.assert.isUndefined (node_modules/chai/lib/chai/interface/assert.js:478:32)
      at test/index.test.js:35:14

  4) top_sites #65 Realtor should be scrapeable by embedly:
     AssertionError: expected 'HTTP 503: Service Unavailable' to equal undefined
      at Function.assert.isUndefined (node_modules/chai/lib/chai/interface/assert.js:478:32)
      at test/index.test.js:35:14
```

--- a/README.md
+++ b/README.md
@@ -10,7 +10,15 @@ $ python make_manifest.py --help
 Usage: make_manifest.py [OPTIONS]
 
 Options:
-  --count INTEGER         Number of sites from Alexa Top Sites
+  --count INTEGER         Number of sites from a list of Top Sites that should
+                          be used to generate the manifest. Default is 10.
+  --topsitesfile PATH     A csv file containing comma separated rank and
+                          domain information (in the same order) of the Top
+                          Sites. If no file is provided then Alexa Top Sites
+                          are used.
+  --minwidth INTEGER      Minimum width of the site icon. Only those sites
+                          that satisfy this requirement are added to the
+                          manifest. Default is 96.
   --loadrawsitedata TEXT  Load the full data from the filename specified
   --saverawsitedata TEXT  Save the full data to the filename specified
   --help                  Show this message and exit.

--- a/make_manifest.py
+++ b/make_manifest.py
@@ -2,6 +2,7 @@ import json
 import logging
 import re
 import zipfile
+import csv
 from io import BytesIO, StringIO
 from urllib.parse import urljoin
 
@@ -49,12 +50,23 @@ def _fetch_alexa_top_sites():
         rank, domain = row.split(',')
         yield (int(rank), domain.strip())
 
+def _fetch_top_sites(topsitesfile):
+    with open(topsitesfile, newline='') as csvfile:
+        rows = csv.reader(csvfile)
+        for row in rows:
+            if len(row) == 0:
+                # skip empty lines
+                continue
+            yield (row[0], row[1])
 
-def alexa_top_sites(count=1000):
-    logging.info('Fetching Alexa top {count} sites'.format(count=count))
-    top = _fetch_alexa_top_sites()
-    return [next(top) for x in range(count)]
-
+def top_sites(topsitesfile, count):
+    logging.info('Fetching top {count} sites'.format(count=count))
+    top_sites_generator = None
+    if topsitesfile:
+        top_sites_generator = _fetch_top_sites(topsitesfile)
+    else:
+        top_sites_generator = _fetch_alexa_top_sites()
+    return [next(top_sites_generator) for x in range(count)]
 
 def fetch_icons(url, user_agent=IPHONE_UA):
     logging.info('Fetching icons for {url}'.format(url=url))
@@ -92,7 +104,7 @@ def fix_url(url):
     return fixed
 
 
-def get_best_icon(images):
+def get_best_icon(minwidth, images):
     image_url = None
     image_width = 0
     for image in images:
@@ -121,16 +133,15 @@ def get_best_icon(images):
             image_url = url
             image_width = width
 
-    if image_width < 96:
-        # We don't want any images under 96px
+    if image_width < minwidth:
+        # We don't want any images under specified resolution
         image_url = None
 
     return image_url
 
-
-def collect_icons_for_alexa_top(count, extra_domains=None):
+def collect_icons_for_top_sites(minwidth, topsitesfile, count, extra_domains=None):
     results = []
-    for rank, hostname in alexa_top_sites(count) + [(-1, x) for x in extra_domains or []]:
+    for rank, hostname in top_sites(topsitesfile, count) + [(-1, x) for x in extra_domains or []]:
         # Skip NSFW and blacklisted sites
         if is_nsfw(hostname) or hostname in DOMAIN_BLACKLIST:
             continue
@@ -141,22 +152,25 @@ def collect_icons_for_alexa_top(count, extra_domains=None):
             # Retry with http
             url = 'http://{hostname}'.format(hostname=hostname)
             icons = fetch_icons(url)
+        best_icon_url = get_best_icon(minwidth, icons)
         results.append({
             'hostname': hostname,
             'url': url,
             'icons': icons,
             'rank': rank,
-            'best_icon': get_best_icon(icons)
+            'best_icon': best_icon_url
         })
     logging.info('Done fetching icons')
     return results
 
 
 @click.command()
-@click.option('--count', default=10, help='Number of sites from Alexa Top Sites')
+@click.option('--count', default=10, help='Number of sites from a list of Top Sites that should be used to generate the manifest. Default is 10.')
+@click.option('--topsitesfile', type=click.Path(exists=True), help='A csv file containing comma separated rank and domain information (in the same order) of the Top Sites. If no file is provided then Alexa Top Sites are used.')
+@click.option('--minwidth', default=96, help='Minimum width of the site icon. Only those sites that satisfy this requirement are added to the manifest. Default is 96.')
 @click.option('--loadrawsitedata', help='Load the full data from the filename specified')
 @click.option('--saverawsitedata', help='Save the full data to the filename specified')
-def make_manifest(count, saverawsitedata, loadrawsitedata):
+def make_manifest(count, minwidth, topsitesfile, saverawsitedata, loadrawsitedata):
     results = []
 
     if loadrawsitedata:
@@ -164,7 +178,7 @@ def make_manifest(count, saverawsitedata, loadrawsitedata):
         with open(loadrawsitedata) as infile:
             sites_with_icons = json.loads(infile.read())
     else:
-        sites_with_icons = collect_icons_for_alexa_top(count, extra_domains=DOMAIN_WHITELIST);
+        sites_with_icons = collect_icons_for_top_sites(minwidth, topsitesfile, count, extra_domains=DOMAIN_WHITELIST);
         if saverawsitedata:
             logging.info('Saving raw icon data to {filename}'.format(filename=saverawsitedata))
             with open(saverawsitedata, 'w') as outfile:

--- a/make_manifest.py
+++ b/make_manifest.py
@@ -60,7 +60,7 @@ def _fetch_top_sites(topsitesfile):
             yield (row[0], row[1])
 
 def top_sites(topsitesfile, count):
-    logging.info('Fetching top {count} sites'.format(count=count))
+    logging.info(f'Fetching top {count} sites')
     top_sites_generator = None
     if topsitesfile:
         top_sites_generator = _fetch_top_sites(topsitesfile)

--- a/make_manifest.py
+++ b/make_manifest.py
@@ -69,7 +69,7 @@ def top_sites(topsitesfile, count):
     return [next(top_sites_generator) for x in range(count)]
 
 def fetch_icons(url, user_agent=IPHONE_UA):
-    logging.info('Fetching icons for {url}'.format(url=url))
+    logging.info(f'Fetching icons for {url}')
     icons = []
     browser = RoboBrowser(user_agent=user_agent, parser='html.parser')
     try:
@@ -174,13 +174,13 @@ def make_manifest(count, minwidth, topsitesfile, saverawsitedata, loadrawsitedat
     results = []
 
     if loadrawsitedata:
-        logging.info('Loading raw icon data from {filename}'.format(filename=loadrawsitedata))
+        logging.info(f'Loading raw icon data from {loadrawsitedata}')
         with open(loadrawsitedata) as infile:
             sites_with_icons = json.loads(infile.read())
     else:
         sites_with_icons = collect_icons_for_top_sites(minwidth, topsitesfile, count, extra_domains=DOMAIN_WHITELIST);
         if saverawsitedata:
-            logging.info('Saving raw icon data to {filename}'.format(filename=saverawsitedata))
+            logging.info(f'Saving raw icon data to {saverawsitedata}')
             with open(saverawsitedata, 'w') as outfile:
                 json.dump(sites_with_icons, outfile, indent=4)
 

